token-swap/program/src/curve/calculator.rs

```diff
@@ -225,4 +238,45 @@ pub mod test {
             deposit_token_b.token_b_amount,
         );
     }
+
+    pub fn check_curve_value_from_swap(
+        curve: &dyn CurveCalculator,
+        source_token_amount: u128,
+        swap_source_amount: u128,
+        swap_destination_amount: u128,
+        trade_direction: TradeDirection,
+    ) {
+        let results = curve
+            .swap_without_fees(
+                source_token_amount,
+                swap_source_amount,
+                swap_destination_amount,
+                trade_direction,
+            )
+            .unwrap();
+
+        let (swap_token_a_amount, swap_token_b_amount) = match trade_direction {
+            TradeDirection::AtoB => (swap_source_amount, swap_destination_amount),
+            TradeDirection::BtoA => (swap_destination_amount, swap_source_amount),
+        };
+        let previous_value = curve
+            .total_value(swap_token_a_amount, swap_token_b_amount)
+            .unwrap();
+
+        let new_swap_source_amount = swap_source_amount
+            .checked_add(results.source_amount_swapped)
+            .unwrap();
+        let new_swap_destination_amount = swap_destination_amount
+            .checked_sub(results.destination_amount_swapped)
+            .unwrap();
+        let (swap_token_a_amount, swap_token_b_amount) = match trade_direction {
+            TradeDirection::AtoB => (new_swap_source_amount, new_swap_destination_amount),
+            TradeDirection::BtoA => (new_swap_destination_amount, new_swap_source_amount),
+        };
+
+        let new_value = curve
+            .total_value(swap_token_a_amount, swap_token_b_amount)
+            .unwrap();
+        assert!(new_value >= previous_value);
```

Comment:
It feels worth adding an error bounds for this test too. We wouldn't want a regression that causes the invariant to bump too much because that would indicate that swaps are too lossy for traders